internal/admission/validation/consumers/credentials/validation.go

```diff
@@ -15,12 +17,19 @@ import (
 // ValidateCredentials performs basic validation on a credential secret given
 // the Kubernetes secret which contains credentials data.
 func ValidateCredentials(secret *corev1.Secret) error {
-	// the indication of credential type is required to be present on all credentials.
-	credentialTypeB, ok := secret.Data[TypeKey]
-	if !ok {
-		return fmt.Errorf("missing required key %s", TypeKey)
+	var credentialType string
+	var labelOk bool
+	credentialType, labelOk = secret.Labels[labels.LabelPrefix+labels.CredentialKey]
+
+	if !labelOk {
+		// the indication of credential type is required to be present on all credentials.
+		credentialTypeB, ok := secret.Data[TypeKey]
+		if !ok {
+			// while it's also missing the data field, encourage the new label convention
+			return fmt.Errorf("credential missing %s label", labels.LabelPrefix+labels.CredentialKey)
+		}
+		credentialType = string(credentialTypeB)
 	}
```

Comment:
Now we have this logic scattered in various forms in 3 places: here, `FillConsumersAndCredentials` and `RequestHandler.handleSecret`. Could we create one function `credentials.ExtractCredentialType(*corev1.Secret) (type string, ok bool)` and use it in all of the places?

Reply:
This step in particular is now separate, though there wasn't a great package to hold consumer logic shared between `admission` and `kongstate`. I felt silly creating a `credentials` package that would _only_ hold this, so I stuck it in `util`, the designated island of misfit code.

The logic for handling credentials is a pretty big mess that could probably be refactored into reuseable components and better separation of concerns between steps, but IDK if I want to try and sort that out for 3.0 given the short deadline we have to complete the remaining scheduled items.

Sorting that out should provide enough shared functions to fill out a `credentials` package, and https://github.com/Kong/kubernetes-ingress-controller/issues/4853 seemed a good follow-up to try and work it in.